### PR TITLE
[RIP-9] The best practice of rocketmq

### PR DESCRIPTION
**Pull request by fusiming**
## What is the purpose of the change

RocketMQ Developer Guide doc

## Brief changelog

Add the file named best_practice.md

## Verifying this change

no functional change

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: docs/cn/best_practise.md [hunk 1]
```diff
@@ -0,0 +1,574 @@
+7 最佳实践(best practice) {#h.wri31g59kesl .c6}
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 请删除无关代码

File: docs/cn/best_practise.md [hunk 2]
```diff
@@ -0,0 +1,574 @@
...
+=========================
+
+7.1 客户端 {#h.ml9txd5nlf7z .c126}
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 请删除无关代码

**Comment by rudy2steiner** [comment]
great

File: docs/cn/best_practise.md [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+
+RocketMQ可以令客户端找到Name Server, 然后通过Name Server再找到Broker，分别如下，如下所示有多种配置方式，优先级由高到低，高优先级会覆盖低优先级。
+
+1. 代码中指定Name Server地址
+   // 多个namesrv地址之间用分号分割   
+
+```java
+producer.setNamesrvAddr("192.168.0.1:9876;192.168.0.2:9876");  
+```
+或
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: “或”delete

File: docs/cn/best_practise.md [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+## 7.1 客户端
+
+​       相对于RocketMQ的Broker集群，生产者和消费者都是客户端。本小节主要描述生产者和消费者公共的行为配置。
+
+### 7.1.1 客户端寻址方式
+
+RocketMQ可以令客户端找到Name Server, 然后通过Name Server再找到Broker，分别如下，如下所示有多种配置方式，优先级由高到低，高优先级会覆盖低优先级。
+
+1. 代码中指定Name Server地址
+   // 多个namesrv地址之间用分号分割
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: “//"建议用","

File: docs/cn/best_practise.md [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+| producerGroup                    | DEFAULT_PRODUCER | Producer组名，多个Producer如果属于一个应用，发送同样的消息，则应该将它们归为同一组 |
+| createTopicKey                   | TBW102           | 在发送消息时，自动创建服务器不存在的topic，需要指定Key，该Key可用于配置发送消息所在topic的默认路由。 |
+| defaultTopicQueueNums            | 4                | 在发送消息，自动创建服务器不存在的topic时，默认创建的队列数  |
+| sendMsgTimeout                   | 10000            | 发送消息超时时间，单位毫秒                                   |
+| compressMsgBodyOverHowmuch       | 4096             | 消息Body超过多大开始压缩（Consumer收到消息会自动解压缩），单位字节 |
+| retryAnotherBrokerWhenNotStoreOK | FALSE            | 如果发送消息返回sendResult，但是sendStatus!=SEND_OK，是否重试发送 |
+| retryTimesWhenSendFailed         | 2                | 如果消息发送失败，最大重试次数，该参数只对同步发送模式起作用 |
+| maxMessageSize                   | 4MB              | 客户端限制的消息大小，超过报错，同时服务端也会限制，所以需要跟服务端配合使用。 |
+| transactionCheckListener         |                  | 事务消息回查监听器，如果发送事务消息，必须设置               |
+| checkThreadPoolMinSize           | 1                | Broker回查Producer事务状态时，线程池大小                     |
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 说明需斟酌

File: docs/cn/best_practise.md [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+| checkThreadPoolMaxSize           | 1                | Broker回查Producer事务状态时，线程池大小                     |
+| checkRequestHoldMax              | 2000             | Broker回查Producer事务状态时，Producer本地缓冲请求队列大小   |
+| RPCHook                          | null             | 该参数是在Producer创建时传入的，包含消息发送前的预处理和消息响应后的处理两个接口，用户可以在第一个接口中做一些安全控制或者其他操作。 |
+
+#### 3  PushConsumer配置
+
+| 参数名                       | 默认值                        | 说明                                                         |
+| ---------------------------- | ----------------------------- | ------------------------------------------------------------ |
+| consumerGroup                | DEFAULT_CONSUMER              | Consumer组名，多个Consumer如果属于一个应用，订阅同样的消息，且消费逻辑一致，则应该将它们归为同一组 |
+| messageModel                 | CLUSTERING                    | 消息模型，支持以下两种   1、集群消费   2、广播消费           |
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: “消费模型支持集群消费和广播消费两种”

File: docs/cn/best_practise.md [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+| consumeFromWhere             | CONSUME_FROM_LAST_OFFSET      | Consumer启动后，默认从上次消费的位置开始消费，这包含两种情况：一种是上次消费的位置未过期，则消费从上次中止的位置进行；一种是上次消费位置已经过期，则从当前队列第一条消息开始消费 |
+| consumeTimestamp             | 半个小时前                    | 只有当consumeFromWhere值为CONSUME_FROM_TIMESTAMP时才起作用。 |
+| allocateMessageQueueStrategy | AllocateMessageQueueAveragely | Rebalance算法实现策略                                        |
+| subscription                 | {}                            | 订阅关系                                                     |
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: {}?

File: docs/cn/best_practise.md [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+| messageListener              |                               | 消息监听器                                                   |
+| offsetStore                  |                               | 消费进度存储                                                 |
+| consumeThreadMin             | 10                            | 消费线程池数量                                               |
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 消费线程池最大线程数

File: docs/cn/best_practise.md [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+
+#### 4  PullConsumer配置
+
+| 参数名                           | 默认值                        | 说明                                                         |
+| -------------------------------- | ----------------------------- | ------------------------------------------------------------ |
+| consumerGroup                    | DEFAULT_CONSUMER              | Consumer组名，多个Consumer如果属于一个应用，订阅同样的消息，且消费逻辑一致，则应该将它们归为同一组 |
+| brokerSuspendMaxTimeMillis       | 20000                         | 长轮询，Consumer拉消息请求在Broker挂起最长时间，单位毫秒     |
+| consumerTimeoutMillisWhenSuspend | 30000                         | 长轮询，Consumer拉消息请求在Broker挂起超过指定时间，客户端认为超时，单位毫秒 |
+| consumerPullTimeoutMillis        | 10000                         | 非长轮询，拉消息超时时间，单位毫秒                           |
+| messageModel                     | BROADCASTING                  | 消息模型，支持以下两种   1、集群消费   2、广播消费           |
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 说明同push consumer

File: docs/cn/best_practise.md [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+| registerTopics                   | []                            | 注册的topic集合                                              |
+| allocateMessageQueueStrategy     | AllocateMessageQueueAveragely | Rebalance算法实现策略                                        |
+
+#### 5  Message数据结构
+
+| 字段名         | 默认值 | 说明                                                         |
+| -------------- | ------ | ------------------------------------------------------------ |
+| Topic          | null   | 必填，线下环境不需要申请，线上环境需要申请后才能使用         |
+| Body           | null   | 必填，字节类型，序列化由应用决定，Producer与Consumer要协商好序列化形式。 |
+| Tags           | null   | 选填，类似于Gmail为每封邮件设置的标签，方便服务器过滤使用。目前只支持每个消息设置一个tag，所以也可以类比为Notify的MessageType概念 |
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: “所以也可以类比为Notify的MessageType概念”删除

File: docs/cn/best_practise.md [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+| -------------------------------- | ----------------------------- | ------------------------------------------------------------ |
+| consumerGroup                    | DEFAULT_CONSUMER              | Consumer组名，多个Consumer如果属于一个应用，订阅同样的消息，且消费逻辑一致，则应该将它们归为同一组 |
+| brokerSuspendMaxTimeMillis       | 20000                         | 长轮询，Consumer拉消息请求在Broker挂起最长时间，单位毫秒     |
+| consumerTimeoutMillisWhenSuspend | 30000                         | 长轮询，Consumer拉消息请求在Broker挂起超过指定时间，客户端认为超时，单位毫秒 |
+| consumerPullTimeoutMillis        | 10000                         | 非长轮询，拉消息超时时间，单位毫秒                           |
+| messageModel                     | BROADCASTING                  | 消息模型，支持以下两种   1、集群消费   2、广播消费           |
+| messageQueueListener             |                               | 监听队列变化                                                 |
+| offsetStore                      |                               | 消费进度存储                                                 |
+| registerTopics                   | []                            | 注册的topic集合                                              |
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: []?

File: docs/cn/best_practise.md [hunk 12]
```diff
@@ -0,0 +1,339 @@
...
+```
+每个消息在业务层面的唯一标识码要设置到keys字段，方便将来定位消息丢失问题。服务器会为每个消息创建索引（哈希索引），应用可以通过topic、key来查询这条消息内容，以及消息被谁消费。由于是哈希索引，请务必保证key尽可能唯一，这样可以避免潜在的哈希冲突。
+```java
+   // 订单Id   
+   String orderId = "20034568923546";   
+   message.setKeys(orderId);   
+```
+​     消息发送成功或者失败要打印消息日志，务必要打印SendResult和key字段。send消息方法只要不抛异常，就代表发送成功。发送成功会有多个状态，在sendResult里定义。以下对每个状态进行说明：
+
+●        SEND_OK（发送成功）
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 用项目编号-

File: docs/cn/best_practise.md [hunk 13]
```diff
@@ -0,0 +1,339 @@
...
+3)      如果本身向broker发送消息产生超时异常，就不会再重试。
+
+​       以上策略也是在一定程度上保证了消息可以发送成功。如果业务对消息可靠性要求比较高，建议应用增加相应的重试逻辑：比如调用send同步方法发送失败时，则尝试将消息存储到db，然后由后台线程定时重试，确保消息一定到达Broker。
+
+​      上述db重试方式为什么没有集成到MQ客户端内部做，而是要求应用自己去完成？我们基于以下几点考虑：首先、MQ的客户端设计为无状态模式，方便任意的水平扩展，且对机器资源的消耗仅仅是cpu、内存、网络。第二、如果MQ客户端内部集成一个KV存储模块，那么数据只有同步落盘才能较可靠，而同步落盘本身性能开销较大，所以通常会采用异步落盘，又由于应用关闭过程不受MQ运维人员控制，可能经常会发生 kill -9 这样暴力方式关闭，造成数据没有及时落盘而丢失。第三、Producer所在机器的可靠性较低，一般为虚拟机，不适合存储重要数据。综上，建议重试过程交由应用来控制。
+
+### 7.2.3选择oneway形式发送
+
+​     通常消息的发送是这样一个过程：
+​	1)      客户端发送请求到服务器
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 文字格式，下同

File: docs/cn/best_practise.md [hunk 14]
```diff
@@ -0,0 +1,339 @@
...
+​	2)      服务器处理请求
+​	3)      服务器向客户端返回应答
+
+​      所以，一次消息发送的耗时时间是上述三个步骤的总和，而某些场景要求耗时非常短，但是对可靠性要求并不高，例如日志收集类应用，此类应用可以采用oneway形式调用，oneway形式只发送请求不等待应答，而发送请求在客户端实现层面仅仅是一个操作系统系统调用的开销，即将数据写入客户端的socket缓冲区，此过程耗时通常在微秒级。
+
+### 7.2.4 其他
+
+1)      建议消息的大小最好不要超过512K。
+
+异步发送（AsyncSending） 默认方法send（msg）会在得到返回响应之前一直被阻塞。因此，如果用户重视性能，那么建议使用send（msg、callback），这样消息发送将以异步方式进行。
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 与标题无关，建议delete

File: docs/cn/best_practise.md [hunk 15]
```diff
@@ -0,0 +1,339 @@
...
+​       RocketMQ无法避免消息重复（Exactly-Once），所以如果业务对消费重复非常敏感，务必要在业务层面进行去重处理。可以借助关系数据库进行去重。首先需要确定消息的唯一键，可以是msgId，也可以是消息内容中的唯一标识字段，例如订单Id等。在消费之前判断唯一键是否在关系数据库中存在。如果不存在则插入，并消费，否则跳过。（实际过程要考虑原子性问题，判断是否存在可以尝试插入，如果报主键冲突，则插入失败，直接跳过）
+
+msgId一定是全局唯一标识符，但是实际使用中，可能会存在相同的消息有两个不同msgId的情况（消费者主动重发、因客户端重投机制导致的重复等），这种情况就需要使业务字段进行重复消费。
+
+### 7.3.2 消费速度慢的处理方式
+
+1   提高消费并行度
+
+​     绝大部分消息消费行为都属于 IO 密集型，即可能是操作数据库，或者调用 RPC，这类消费行为的消费速度在于后端数据库或者外系统的吞吐量，通过增加消费并行度，可以提高总的消费吞吐量，但是并行度增加到一定程度，反而会下降。所以，应用必须要设置合理的并行度。 如下有几种修改消费并行度的方法：
+​	1)  同一个 ConsumerGroup 下，通过增加 Consumer 实例数量来提高并行度（需要注意的是超过订阅队列数的 Consumer 实例无效）。可以通过加机器，或者在已有机器启动多个进程的方式。
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 项目编号-

File: docs/cn/best_practise.md [hunk 16]
```diff
@@ -0,0 +1,339 @@
...
+            // TODO 消息堆积情况的特殊处理
+            return ConsumeConcurrentlyStatus.CONSUME_SUCCESS;
+        }
+        // TODO 正常消费过程
+        return ConsumeConcurrentlyStatus.CONSUME_SUCCESS;
+    }    
+```
+4   优化每条消息消费过程      
+举例如下，某条消息的消费过程如下：
+    1) 根据消息从 DB 查询【数据 1】
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 格式

File: docs/cn/best_practise.md [hunk 17]
```diff
@@ -0,0 +1,339 @@
...
+    3) 复杂的业务计算
+    4) 向 DB 插入【数据 3】
+    5) 向 DB 插入【数据 4】
+
+​      这条消息的消费过程中有4次与 DB的 交互，如果按照每次 5ms 计算，那么总共耗时 20ms，假设业务计算耗时 5ms，那么总过耗时 25ms，所以如果能把 4 次 DB 交互优化为 2 次，那么总耗时就可以优化到 15ms，即总体性能提高了 40%。所以应用如果对时延敏感的话，可以把DB部署在SSD硬盘，相比于SCS磁盘，前者的RT会小很多。
+
+### 7.3.3 消费打印日志
+
+如果消息量较少，建议在消费入口方法打印消息，消费耗时等，方便后续排查问题。
+​```java
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 格式

File: docs/cn/best_practise.md [hunk 18]
```diff
@@ -0,0 +1,339 @@
...
+
+​    不建议阻塞监听器，因为它会阻塞线程池，并最终可能会终止消费进程
+
+6 关于线程数设置     
+
+​    消费者使用 ThreadPoolExecutor 在内部对消息进行消费，所以你可以通过设置 setConsumeThreadMin 或 setConsumeThreadMax 来改变它。
+
+7 关于消费位点
+
+​    当建立一个新的消费者组时，需要决定是否需要消费已经存在于 Broker 中的历史消息CONSUME_FROM_LAST_OFFSET 将会忽略历史消息，并消费之后生成的任何消息。CONSUME_FROM_FIRST_OFFSET 将会消费每个存在于 Broker 中的信息。你也可以使用 CONSUME_FROM_TIMESTAMP 来消费在指定时间戳后产生的消息。
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: "...于Broker中的历史消息”后面是句号

File: docs/cn/best_practise.md [hunk 19]
```diff
@@ -0,0 +1,339 @@
...
+
+## 7.6  系统配置
+
+​      本小节主要介绍系统（JVM/OS）相关的配置。
+
+### 7.6.1 JVM选项
+
+​ 推荐使用最新发布的JDK 1.8版本。通过设置相同的Xms和Xmx值来防止JVM调整堆大小以获得更好的性能。简单的JVM配置如下所示：
+​  -server -Xms8g -Xmx8g -Xmn4g   
+​  如果您不关心RocketMQ代理的启动时间，还有一种更好的选择，就是通过“预触摸”Java堆以确保在JVM初始化期间每个页面都将被分配。那些不关心启动时间的人可以启用它：
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: “RocketMQ 代理”改为Broker

File: docs/cn/best_practise.md [hunk 20]
```diff
@@ -0,0 +1,339 @@
...
+​ -XX:+AlwaysPreTouch   
+禁用偏置锁定可能会减少JVM暂停，
+​ -XX:-UseBiasedLocking   
+至于垃圾回收，建议使用带JDK 1.8的G1收集器。
+```text
+-XX:+UseG1GC -XX:G1HeapRegionSize=16m   
+-XX:G1ReservePercent=25 
+-XX:InitiatingHeapOccupancyPercent=30
+```
+​ 这些GC选项看起来有点激进，但事实证明它在我们的生产环境中具有良好的性能。另外不要把-XX:MaxGCPauseMillis的值设置太小，否则JVM将使用一个小的年轻代来实现这个目标，这将导致非常频繁的minor GC，所以官方建议使用rolling GC日志文件：
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: “官方”delete

File: docs/cn/best_practise.md [hunk 21]
```diff
@@ -0,0 +1,339 @@
...
+```
+如果写入GC文件会增加代理的延迟，可以考虑将GC日志文件重定向到内存文件系统：
+```text
+-Xloggc:/dev/shm/mq_gc_%p.log123   
+```
+### 7.6.2 Linux内核参数
+
+​ os.sh脚本在bin文件夹中列出了许多内核参数，可以进行微小的更改然后用于生产用途。下面的参数需要注意，更多细节请参考/proc/sys/vm/*的文档。
+（<https://www.kernel.org/doc/Documentation/sysctl/vm.txt> ）
+​ vm.extra_free_kbytes，告诉VM在后台回收（kswapd）启动的阈值与直接回收（通过分配进程）的阈值之间保留额外的可用内存.RocketMQ使用此参数来避免内存分配中的长延迟。（与具体内核版本相关）
```
Comment (wlliqipeng) [inline docs/cn/best_practise.md]: 格式

**Pull request closed without merging**